app/models/concerns/hostext/fact_data.rb

```diff
@@ -0,0 +1,58 @@
+module Hostext
+  module FactData
+    extend ActiveSupport::Concern
+
+    def set_reported_data(parser)
+      set_data(:boot_time, :boot_timestamp, parser, ->(value) { Time.at value })
+      set_data(:virtual, :virtual, parser)
+      set_data(:ram, :ram, parser, ->(value) { format_host_memory value })
+      set_data(:sockets, :sockets, parser)
+      set_data(:cores, :cores, parser)
+    end
+
+    def cores
+      get_data :cores
+    end
+
+    def virtual
+      get_data :virtual
+    end
+
+    def sockets
+      get_data :sockets
+    end
+
+    def ram
+      get_data :ram
+    end
+
+    def uptime_seconds
+      get_data(:boot_time, ->(value) { value.nil? ? nil : Time.zone.now.to_i - value.to_i })
+    end
+
+    def reported_data_facet
+      self.reported_data || self.build_reported_data
+    end
+
+    private
+
+    def get_data(attr_name, format_lambda = ->(value) { value })
+      value = self&.reported_data&.public_send(attr_name)
+      format_lambda.call(value)
+    end
+
+    def set_data(attr_name, method_name, parser, format_lambda = ->(value) { value })
+      value = parser.public_send(method_name)
+      return unless value
+
+      reported_data_facet.update(attr_name => format_lambda.call(value)) if self.persisted?
+    end
+
+    def format_host_memory(memory)
+      return (memory / 1048576).round(2) unless memory.is_a? String
```

Comment:
for facter facts, we could use `to_gb` from our core extensions https://github.com/theforeman/foreman/blob/develop/lib/core_extensions.rb#L139-L156

if we get value as a number in bytes, it also works, we just need to convert it to string

Reply:
also note, subscription manager reports in megabytes, if we use to_gb, the implementation in katello needs to * 1024